src/object/obj_tx.c

```diff
@@ -2200,13 +2203,34 @@ dc_tx_commit_trigger(tse_task_t *task, struct dc_tx *tx, daos_tx_commit_t *args)
 	crt_req_decref(req);
 	crt_req_decref(req);
 out:
-	if (rc == -DER_TX_RESTART)
-		tx->tx_status = TX_FAILED;
-	else if (rc != 0)
-		tx->tx_status = TX_ABORTED;
+	if (rc == -DER_STALE) {
+		dcsr = &tx->tx_req_cache[dc_tx_leftmost_req(tx, false)];
+		rc = obj_pool_query_task(tse_task2sched(task), dcsr->dcsr_obj, 0, &pool_task);
+		if (rc != 0) {
+			/* It is unnecessary to restart the TX if failed to
+			 * refresh pool map to avoid falling into deap loop.
+			 */
+			D_ERROR("Failed to refresh pool map for "DF_DTI": "DF_RC"\n",
+				DP_DTI(&tx->tx_id), DP_RC(rc));
+		} else {
+			rc = -DER_TX_RESTART;
+			tx->tx_status = TX_FAILED;
+			/* Unlock for dc_tx_commit held. */
+			D_MUTEX_UNLOCK(&tx->tx_lock);
+			locked = false;
+			tse_task_schedule(pool_task, true);
```

Comment:
need not set dependency between task and pool_task? do you need to ensure pool map refresh finished before re-execute task?

Reply:
Normally, if the original task will be retried automatically, it needs to depend on the pool_task. But for DTX case, we will fail current TX via DER_TX_RESTART. As for how the upper later caller to handle DER_TX_RETART, that depends on the caller, it may restart the TX with re-executing related operations and then commit the new TX (with new epoch), or it can fail the TX directly. So here, we kickoff the pool task directly. If it is completed before TX restarted, then it is fine; otherwise, it may cause another pool query task if the TX restarted so quickly, but that is not fatal.